packages/runtime/src/state/modules/element-trees.ts

```diff
@@ -0,0 +1,358 @@
+import { type Operation } from 'ot-json0'
+import { getIn } from 'immutable'
+
+import { type Element, ElementData, isElementReference } from '@makeswift/controls'
+
+import { type Action, ActionTypes } from '../actions'
+import { Introspection } from '../../prop-controllers'
+
+import { type DescriptorsByComponentType } from './prop-controllers'
+
+export type ElementTree = {
+  elements: Map<string, Element>
+  elementIds: Map<string, string>
+}
+
+export type State = Map<string, ElementTree>
+
+export function getInitialState(
+  rootElements?: Map<string, Element>,
+  descriptors?: DescriptorsByComponentType,
+): State {
+  const state = new Map<string, ElementTree>()
+  if (rootElements == null || descriptors == null) return state
+
+  for (const [documentKey, rootElement] of rootElements) {
+    state.set(documentKey, buildElementTree(rootElement, descriptors))
+  }
+
+  return state
+}
+
+function getElementTree(state: State, id: string): ElementTree | null {
+  return state.get(id) ?? null
+}
+
+export function getElements(state: State, documentKey: string): Map<string, Element> {
+  return getElementTree(state, documentKey)?.elements ?? new Map()
```

Comment:
This `?? new Map()` will return a new reference each time this function is called so unless we memoize it it could cause unnecessary re-renders